Osdocs 8400 migrate open shift concepts (#75415)
openshift/openshift-docs: cloud_experts_tutorials/cloud-experts-getting-started/cloud-experts-getting-started-openshift-concepts.adoc

For example, to create a reproducible build pipeline for Tomcat and Maven:

. Create a builder image containing OpenJDK and Tomcat that expects to have a WAR file injected.

. Create a second image that layers on top of the first image Maven and any other standard dependencies, and expects to have a Maven project injected.

. Start S2I using the Java application source and the Maven image to create the desired application WAR.

. Start S2I a second time using the WAR file from the earlier step and the initial Tomcat image to create the runtime image.

By placing build logic inside of images and combining the images into multiple steps, the runtime environment is close to the build environment without requiring the deployment of build tools to production.

=== S2I benefits

> ==== Reproducibility
> Allow build environments to be tightly versioned by encapsulating them within a container image and defining a simple interface of injected source code for callers. Reproducible builds are a key requirement for enabling security updates and continuous integration in containerized infrastructure, and builder images help ensure repeatability and the ability to swap run times.
> 
> ==== Flexibility
> Any existing build system that can run on Linux can run inside of a container, and each individual builder can also be part of a larger pipeline. The scripts that process the application source code can be injected into the builder image, allowing authors to adapt existing images to enable source handling.
> 
> ==== Speed
> Instead of building multiple layers in a single Dockerfile, S2I encourages authors to represent an application in a single image layer. This saves time during creation and deployment and allows for better control over the output of the final image.
> 
> ==== Security
> Dockerfiles are run without many of the normal operational controls of containers. They usually run as root and have access to the container network. S2I can control what permissions and privileges are available to the builder image since the build is launched in a single container. In concert with platforms like OpenShift, S2I allows administrators to control what privileges developers have at build time.

Reviewer: Maybe these individual "benefits" would work better as a description list instead of little subsections?
```suggestion
Reproducibility::
Allow build environments to be tightly versioned by encapsulating them within a container image and defining a simple interface of injected source code for callers. Reproducible builds are a key requirement for enabling security updates and continuous integration in containerized infrastructure, and builder images help ensure repeatability and the ability to swap run times.

Flexibility::
Any existing build system that can run on Linux can run inside of a container, and each individual builder can also be part of a larger pipeline. The scripts that process the application source code can be injected into the builder image, allowing authors to adapt existing images to enable source handling.

Speed::
Instead of building multiple layers in a single Dockerfile, S2I encourages authors to represent an application in a single image layer. This saves time during creation and deployment and allows for better control over the output of the final image.

Security::
Dockerfiles are run without many of the normal operational controls of containers. They usually run as root and have access to the container network. S2I can control what permissions and privileges are available to the builder image since the build is launched in a single container. In concert with platforms like OpenShift, S2I allows administrators to control what privileges developers have at build time.
```

== Routes

> An OpenShift Route exposes a service at a hostname so that external clients can reach it by name. When a Route object is created on OpenShift, it gets picked up by the built-in HAProxy load balancer to expose the requested service and make it externally available with the given configuration. Similar to the Kubernetes Ingress object, Red Hat created the concept of Route to fill a need and then contributed the design principles behind it to the community, which heavily influenced the Ingress design. A Route does have some additional features as can be seen in the following chart.

Reviewer: * Consider [repo guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#api-object-formatting) on API object references (general vs object) re: "route" and "ingress"
* Maybe break up the para
* End with colon
```suggestion
An OpenShift route exposes a service at a hostname so that external clients can reach it by name. When a `Route` object is created on OpenShift, it gets picked up by the built-in HAProxy load balancer to expose the requested service and make it externally available with the given configuration.

Similar to the Kubernetes `Ingress` object, Red Hat created the concept of a route to fill a need and then contributed the design principles behind it to the community, which heavily influenced the `Ingress` object design. A route does have some additional features, as can be seen in the following chart:
```

image::ostoy-routes_ingress_comparison.png[Comparing routes and ingresses]

[NOTE]
====
DNS resolution for a hostname is handled separately from routing. Your administrator might have configured a cloud domain that will always correctly resolve to the router or modify your unrelated hostname DNS records independently to resolve to the router.
====

> An individual route can override some defaults by providing specific configurations in its annotations. See xref:../../networking/routes/route-configuration.adoc#nw-route-specific-annotations_route-configuration[route specific annotations] for more details.

Reviewer: Nit on sentence form per [repo guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#lead-in-link-sentence-style-for-non-xref-situations):
```suggestion
An individual route can override some defaults by providing specific configurations in its annotations. For more information, see xref:../../networking/routes/route-configuration.adoc#nw-route-specific-annotations_route-configuration[Route-specific annotations].
```
or maybe better:
```suggestion
An individual route can override some defaults by providing specific configurations in its annotations.

[role="_additional-resources"]
.Additional resources
* xref:../../networking/routes/route-configuration.adoc#nw-route-specific-annotations_route-configuration[Route-specific annotations]
```

== ImageStreams
An ImageStream stores a mapping of tags to images, metadata overrides that are applied when images are tagged in a stream, and an optional reference to a Docker image repository on a registry.

=== ImageStream benefits

> Using an ImageStream makes it easy to change a tag for a container image. Otherwise to change a tag you need to download the whole image, change it locally, then push it all back. Also promoting applications by having to do that to change the tag and then update the deployment object entails many steps. With ImageStreams you upload a container image once and then you manage its virtual tags internally in OpenShift. In one project you might use the developer tag and only change a reference to it internally, while in production you might use a production tag and also manage it internally. You do not have to deal with the registry.

Reviewer: Various suggestions:
```suggestion
Using an image stream makes it easier to change a tag for a container image. Otherwise, to manually change a tag, you must download the image, change it locally, then push it back. Promoting applications by manually changing a tag and updating the deployment object entails many steps.

With image streams, you upload a container image once, and then you manage its virtual tags internally in OpenShift. In one project you might use a development tag and only change a reference to it internally, while in production you might use a production tag and also manage it internally. You do not have to deal with the registry.
```

> You can also use ImageStreams in conjunction with DeploymentConfigs to set a trigger that will start a deployment as soon as a new image appears or a tag changes its reference.

Reviewer: For all of this, consider the [repo guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#api-object-formatting) on API object references (general vs object) re: "image streams" and "deployment configs".

> See below for more details:
> 
> * link:https://blog.openshift.com/image-streams-faq/[Image Streams Blog post]

Reviewer: ```suggestion
[role="_additional-resources"]
.Additional resources
* link:https://blog.openshift.com/image-streams-faq/[Red Hat Blog: How to Simplify Container Image Management in Kubernetes with OpenShift Image Streams]
```

== Builds
A build is the process of transforming input parameters into a resulting object. Most often, the process is used to transform input parameters or source code into a runnable image. A BuildConfig object is the definition of the entire build process.

OpenShift Container Platform leverages Kubernetes by creating Docker-formatted containers from build images and pushing them to a container image registry.

> Build objects share common characteristics: inputs for a build, the need to complete a build process, logging the build process, publishing resources from successful builds, and publishing the final status of the build. Builds take advantage of resource restrictions, specifying limitations on resources such as CPU usage, memory usage, and build or pod execution time.

Reviewer: Maybe a list for this?
```suggestion
Build objects share common characteristics:

* Inputs for a build
* Requirements to complete a build process
* Logging the build process
* Publishing resources from successful builds
* Publishing the final status of the build

Builds take advantage of resource restrictions, specifying limitations on resources such as CPU usage, memory usage, and build or pod execution time.
```

> See link:https://docs.openshift.com/container-platform/latest/cicd/builds/understanding-image-builds.html[Understanding image builds] for more details.

Reviewer: ```suggestion
[role="_additional-resources"]
.Additional resources
* link:https://docs.openshift.com/container-platform/latest/cicd/builds/understanding-image-builds.html[Understanding image builds]
```

==== Flexibility
Any existing build system that can run on Linux can run inside of a container, and each individual builder can also be part of a larger pipeline. The scripts that process the application source code can be injected into the builder image, allowing authors to adapt existing images to enable source handling.

==== Speed
Instead of building multiple layers in a single Dockerfile, S2I encourages authors to represent an application in a single image layer. This saves time during creation and deployment and allows for better control over the output of the final image.

==== Security
Dockerfiles are run without many of the normal operational controls of containers. They usually run as root and have access to the container network. S2I can control what permissions and privileges are available to the builder image since the build is launched in a single container. In concert with platforms like OpenShift, S2I allows administrators to control what privileges developers have at build time.

> == Routes

Reviewer 2: "Route" part looks good. thanks
